### PR TITLE
Switch Ergodox Infinity over to split_common

### PR DESCRIPTION
**Pull request by firetech**
## Description

Serial Link is cool and all, but since the Infinity is the only keyboard using it, it gets easily forgotten when new split sync features are added.

I had some weird behaviour in that calling the suspend handler for LED Matrix on the slave killed the serial communication, so I added `is_keyboard_master()` to its if statement (which fixed that issue).

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* Somewhat related to #12450

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by firetech** [comment]
Note that the visualizer (not used by default since #13165 was just merged) isn't synced between halves after this. I have [some code to fix this](https://github.com/firetech/qmk_firmware/commit/ed24e019835f62b1f67e9325f9078cb732aee589), that could be added to this PR if desired. Since the visualizer isn't enabled by default, I left it out for now.

File: keyboards/ergodox_infinity/rules.mk
Comment (fauxpark) [inline keyboards/ergodox_infinity/rules.mk:39]: I think `led.c` can be removed altogether, it doesn't look like it does anything.

Comment (firetech) [reply keyboards/ergodox_infinity/rules.mk:39]: True, it's just a placeholder. Removing it.

**Review by drashna — APPROVED** [review]

File: keyboards/ergodox_infinity/config.h
Comment (drashna) [inline keyboards/ergodox_infinity/config.h:58]: This is probably because it's based on the ergodox (OG), which has it's matrix rotated 90 degrees.  So it's probably the same here. 

That said... if that's the case, it may be worth using the eager per row debounce algo for the board, if each row can only be used by one finger at a time.

Comment (firetech) [reply keyboards/ergodox_infinity/config.h:58]: What I referred to here was the fact that the rows (as taken from the old custom matrix.c) are _named_ columns in the schematic, and vice versa. I left it in this state to avoid having to rewrite all the LAYOUT defines.

I've actually tried using the eager_pr debounce algo in the past (I did [some work](https://github.com/firetech/qmk_firmware/commit/a939debc0e0f4d43d34fa82810c5db34f4b3609a) last year to move over the Infinity to the common debounce handling, but ended up not submitting it), mostly for parity with the EZ I use at work, and actually got quite a lot of bouncing keys with that, so I went back to defer_g (which is the main reason I never submitted a PR, it didn't really add any value).

I'm currently using the defer_pk algo on my Infinity, for no particular reason other than having lots of RAM to spare. 😎

File: quantum/led_matrix/led_matrix.c
Comment (fauxpark) [inline quantum/led_matrix/led_matrix.c:462]: Hm, why does LED Matrix need this but RGB Matrix not? Could it be something in the ISSI driver itself?

**Review by fauxpark — APPROVED** [review]

File: quantum/led_matrix/led_matrix.c
Comment (firetech) [reply quantum/led_matrix/led_matrix.c:462]: Not sure. Looking at the code of `led_matrix_set_value_all`, it seems it _shouldn't_ make any difference, but it does. 😕

Comment (firetech) [reply quantum/led_matrix/led_matrix.c:462]: I've done some more testing, and this is really weird.

If I skip the special handling of `LED_MATRIX_SPLIT` in either `led_matrix_set_value_all` or `led_matrix_set_value`, everything works fine, for Ergodox Infinity at least (since its LED pinouts are the same on each half, just mirrored positions). The weird part of this is that the first and second halves of `g_is31_leds` are identical, so the only real difference _should_ be that it's writing the led values twice. 😵 

Note that the serial comms die after suspend regardless of which side is the master (without any change).

**Comment by firetech** [comment]
Forgot to mention that this change makes it "impossible" to use the keyboard halves separately (something that works fine with Serial Link), due to several serial read timeouts each scan cycle eating up the matrix scan rate.

It's possible to solve this by having the split_common system detect that the slave half is missing and holding off on connection attempts for a while (I'm running a [hacky version](https://github.com/firetech/qmk_firmware/commit/b186d532ed814b9a4598a4e0b8d6fafda2d43a38) of this on my keyboard without issues), but I feel that is a bit out of scope for this PR.

**Comment by firetech** [comment]
@drashna This PR on top of develop doesn't really boot properly on my Ergodox Infinity anymore, since #13330 was merged. It hangs during boot and Windows pops up a notification about not being able to identify the last connected USB device. Reverting the #13330 commit makes it work. Any idea on what could be wrong/missing?

**Comment by drashna** [comment]
Thanks!

**Comment by roygbyte** [comment]
Hi. Relative newbie here. I use an Ergodox Infinity and switched from kiibohd to QMK in May 2021. At the time, I used QMK firmware 0.12.50 to build/flash my keyboard. 

The other day I had the bright idea of making some updates/improvements to my configuration and decided to use the new 0.15 firmware to flash. As the story goes, I have noticed a significant reduction in performance and reliability, including:

- Slave half stops receiving input after an unpredictable amount of time
- Random keystrokes seem to be skipped/ignored/etc.

I would like to debug this issue and have figured this PR might be the best place to drop a note, since the switch to `split_common` occurred soon after (i.e.: July 2021) my initial flash in May 2021. I wonder if this is the cause of the issue I am facing. I also wonder if other folks have noticed any performance differences since the switch.

From what I can tell so far, the issue is not with the keyboard configuration itself. I have tested @firetech's lovely configuration on QMK firmware 0.12.50 and 0.15. It works fine with 0.12.50, but experiences the aforementioned issues on 0.15.

**Comment by firetech** [comment]
@roygbyte As you may have noticed in my repository, I haven't updated my build in quite a long while, and my keyboard still runs the exact code in my repository (based on the develop branch back in August)... Life has simply gotten in the way.

I have no idea what has been changed in the recent months, and I'm not sure if the maintainers will see your comment here since this PR has been merged and closed. I recommend checking out the (official) [QMK Discord](https://discord.gg/Uq7gcHh), the people there are usually very helpful. :)

**Comment by roygbyte** [comment]
Thanks for the speedy reply! Yes, I did notice your repo was a bit behind... that could be a good thing, though, if it works as you seem to imply. Perhaps I'll start there and see how that version of QMK performs on my keyboard. (Funny enough, I had been saying to myself the other day that I wanted to work with C code more... low-and-behold, it seems like my wish has been granted xD ) ... or I'll just settle for what works... the cat is a happy addition to my configuration and maybe I should leave it at that! (Those LED effects are so cool, though... if I could get them working too then I'd really be on cloud nine...)

**Comment by firetech** [comment]
@roygbyte I just happened to be browsing GitHub for entirely unrelated reasons and stumbled upon the notification about your post before I noticed the email about it. xD

Actually, since I had some time over just now, I tried rebasing my branch onto current master, and I don't have any issues (except for two core bugfix PRs that I just posted). Everything seems to work fine and dandy here*. My main branch has been updated accordingly (including said PRs).

*) I've been using a build based on master since shortly after my last post. In other words, for about two hours, including hunting bugs and fixing them. In that time, I haven't noticed any of the problems you described.

**Comment by roygbyte** [comment]
Such happenstance to have you stumbled upon my error :)

After a bit of poking around with versions, it seems that between 0.13.38 and 0.14.0 is where my performance degrades. I tested the same layout on both, and while it runs really snappy on the former, it's quite terrible on the later. I should mention that I flashed both using the older instructions for flashing, where the master side is explicitly indicated. I.e.: `make ergodox_infinity:layout MASTER=left`. 

Is it possible there's some other variable causing the issue? I am using USB 3.0 cables, best I could buy. I have the halves connected together... Not sure what else could be causing different results for me vs. you! 

Anyways, I don't mean to take up too much of your time. I have made a bit of discovery, which is good. Thanks for pushing your old version. I'll look around... perhaps there are more clues to reveal...

**Comment by firetech** [comment]
> After a bit of poking around with versions, it seems that between 0.13.38 and 0.14.0 is where my performance degrades.

`git bisect` might help you narrow down the exact commit when performance degrades (it does a binary search between two given revisions). :)

**Comment by roygbyte** [comment]
Thanks for the suggestion. I would not have thought to use that command! If I find anything interesting you might hear from me again 😅

**Comment by roygbyte** [comment]
Which side do you used as the master keyboard? I don't know if/why it would matter, but I seem to get fewer issues when the right side is used as the master. 

I flashed the keyboards with their respective dfu commands:
`make ergodox_infinity:firetech:dfu-util-split-right` for the right side, `make ergodox_infinity:firetech:dfu-util-split-left` for the left side. After this, with the left side as master, the right side will eventually (after a minute or so) stop receiving/transmitting keystrokes. If I switch to the right side as master, I don't get this problem. But: some keys will "lag" and not transmit a keystroke until pressed numerous times.

**Comment by firetech** [comment]
> Which side do you used as the master keyboard? I don't know if/why it would matter, but I seem to get fewer issues when the right side is used as the master.

I mainly use the right side as the master. However, while implementing this PR (and some others), I was switching the master role back and forth, as it made the flashing process easier (flash the master, connect it to the slave, then connect the slave as new master and flash it), so I doubt the direction should matter. Have you tried switching your _cables_ around to see if that makes any difference?

> I flashed the keyboards with their respective dfu commands: `make ergodox_infinity:firetech:dfu-util-split-right` for the right side, `make ergodox_infinity:firetech:dfu-util-split-left` for the left side. After this, with the left side as master, the right side will eventually (after a minute or so) stop receiving/transmitting keystrokes. If I switch to the right side as master, I don't get this problem. But: some keys will "lag" and not transmit a keystroke until pressed numerous times.

I don't have any such issues, weird. Either the hardware, cables or compiler is to blame, I guess. :stuck_out_tongue:

**Comment by roygbyte** [comment]
The break point for me is here: https://github.com/qmk/qmk_firmware/commit/6fe3943ad3efd480da87884ab92fc5738d99f1d. Why that is so I do not know. And, quite frankly, I'm beginning not to care xD

**Pull request merged**
